Docs: update roadmap with integrations (Kafka, Trino, Delta lake as discussion item) (#2359)
treeverse/lakeFS: docs/understand/roadmap.md

[Track and discuss on GitHub](https://github.com/treeverse/lakeFS/issues/1846){: target="_blank" class="btn" }

### Support Delta Lake merges and diffs across branches <span>Requires Discussion</span>{: .label .label-yellow }

New data formats ([Apache Hudi](https://hudi.apache.org/){: target="_blank" }, [Apache Iceberg](https://iceberg.apache.org/){: target="_blank" } and most notably, [Delta Lake](https://delta.io/){: target="_blank" }) don't rely simply on a hierarchical directory structure - they are usually accompanied by metadata files.
These files contain information about changes made, partition and indexing information, as well as represent small deltas to be applied to the larger, typically columnar data objects.

> 

Author: Was actually aiming for _changes made to one more more of (data itself, schema, partitioning, indexing)_ which I can't seem to be able to phrase properly :)

> 

Author: going to merge for now, can always improve phrasing here in a later PR :) Thanks!

For Delta Lake in particular, these metadata files represent a [logical transaction log that relies on numerical ordering](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#delta-log-entries).

Currently, if trying to modify a Delta table from 2 different branches, lakeFS would correctly recognize that there's a conflict: this log diverged into 2 different copies, representing different changes.

> Users would then have to forgo one of the change sets, by either retaining the destination branch's set of changes, or the source branch's.

Reviewer: ```suggestion
Users would then have to forgo one of the change sets, by either retaining the destination's branch set of changes, or the source's branch.
```

> A much better user experience would be to allow merging this log into a new, unified set of changes, representing changes made in both merges, as a new set of log files (and potentially, data files too!).

Reviewer: ```suggestion
A much better user experience would be to allow merging this log into a new unified set of changes, representing changes made in both branches as a new set of log files (and potentially, data files too!).
```

[Track and discuss on GitHub](https://github.com/treeverse/lakeFS/issues/2042){: target="_blank" class="btn" }

### Native connector: Trino

Currently, the Trino integration works well using the [lakeFS s3 Gateway](architecture.md#s3-gateway). 

While easy to integrate and useful out of the box, due to the S3 protocol, it means that the data itself must pass through the lakeFS server.

For larger installations, a native integration where lakeFS handles metadata, returning locations in the underlying object store that Trino can then access directly, would allow reducing the operational overhead and increasing the scalability of lakeFS.
This would be done in a similar way to the [Native Spark integration](../integrations/spark.md) using the [Hadoop Filesystem implementation](../integrations/spark.md#access-lakefs-using-the-lakefs-specific-hadoop-filesystem).

[Track and discuss on GitHub](https://github.com/treeverse/lakeFS/issues/2357){: target="_blank" class="btn" }

### Improved streaming support for Apache Kafka

> Committing (along with attaching useful information on the commit) makes a lot of sense for batch workloads: run a job or a pipeline on a separate branch, and then commit once the pipeline has completed successfully, recording metadata such as the git hash of the code executed, the versions of frameworks used, and information about the data artifacts generated by the run.

Reviewer: This sentence is a bit long and confusing. maybe something like this: 

```suggestion
Committing (along with attaching useful information to the commit) makes a lot of sense for batch workloads: 
- run a job or a pipeline on a separate branch and commit 
- record information such as the git hash of the code executed, the versions of frameworks used, and information about the data artifacts
- once the pipeline has completed successfully, commit and attach the recorded information as metadata

```

For streaming however, this is currently less clear: There's no obvious point in time to commit, as things never actually "finish successfully".
[The recommended pattern](../usecases/production.md#example-1-rollback---data-ingested-from-a-kafka-stream) would be to ingest from a stream on a separate branch, periodically committing - storing not only the data added since last commit but also capturing the offset read from the stream, for reproducibility.
These commits can then be merged into a main branch given they pass all relevant quality checks and other validations using hooks, exposing consumers to validated, clean data.

In practice, implementing such a workflow is a little challenging. Users need to:

1. Orchestrate the commits and merge operations.
2. figure out how to attach the correct offset read from the stream broker
3. Handle writes coming in while the commit is taking place

> Ideally, lakeFS should provide the tools to automate this, with native support for [Apache Kafka](https://kafka.apache.org/){: target="_blank" }.

Reviewer: ```suggestion
Ideally, lakeFS should provide tools to automate this, with native support for [Apache Kafka](https://kafka.apache.org/){: target="_blank" }.
```
